# Exercise 02 -  OLAP Cubes - Grouping Sets

All the databases table in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design is based on O'Reilly's public dimensional modelling tutorial schema [Link](http://archive.oreilly.com/oreillyschool/courses/dba3/index.html)

Start by connecting to the database by running the cells below. If you are coming back to this exercise, then uncomment and run the first cell to recreate the database. If you recently completed the slicing and dicing exercise, then skip to the second cell.

In [ ]:
# !PGPASSWORD=student createdb -h 127.0.0.1 -U student pagila_star
# !PGPASSWORD=student psql -q -h 127.0.0.1 -U student -d pagila_star -f Data/pagila-star.sql

### Connect to the local database where Pagila is loaded

In [1]:
import sql
%load_ext sql

DB_ENDPOINT = "127.0.0.1"
DB = 'pagila_star'
DB_USER = 'student'
DB_PASSWORD = 'student'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)
%sql $conn_string

postgresql://student:student@127.0.0.1:5432/pagila_star


'Connected: student@pagila_star'

### Star Schema

<img src="pagila-star.png" width="50%"/>

# Grouping Sets
- It happens often that for 3 dimensions, you want to aggregate a fact:
    - by nothing (total)
    - then by the 1st dimension
    - then by the 2nd 
    - then by the 3rd 
    - then by the 1st and 2nd
    - then by the 2nd and 3rd
    - then by the 1st and 3rd
    - then by the 1st and 2nd and 3rd
    
- Since this is very common, and in all cases, we are iterating through all the fact table anyhow, there is a more clever way to do that using the SQL grouping statement "GROUPING SETS" 

## Total Revenue

TODO: Write a query that calculates total revenue (sales_amount)

In [2]:
%%sql
SELECT sum(sales_amount) as revenue
FROM factsales

 * postgresql://student:***@127.0.0.1:5432/pagila_star
1 rows affected.


[(Decimal('67416.51'),)]

## Revenue by Country
TODO: Write a query that calculates total revenue (sales_amount) by country

In [10]:
%%sql
SELECT sum(sales_amount) as revenue, country
FROM factsales f join dimcustomer d on f.customer_key=d.customer_key group by country;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
108 rows affected.


[(Decimal('191.47'), 'Cambodia'),
 (Decimal('1662.12'), 'Turkey'),
 (Decimal('831.04'), 'Germany'),
 (Decimal('135.68'), 'Chad'),
 (Decimal('93.78'), 'Madagascar'),
 (Decimal('374.04'), 'France'),
 (Decimal('111.73'), 'Iraq'),
 (Decimal('709.41'), 'Colombia'),
 (Decimal('3471.74'), 'Japan'),
 (Decimal('71.80'), 'American Samoa'),
 (Decimal('746.28'), 'Vietnam'),
 (Decimal('122.68'), 'Virgin Islands, U.S.'),
 (Decimal('111.74'), 'Kuwait'),
 (Decimal('390.13'), 'Ecuador'),
 (Decimal('200.46'), 'Cameroon'),
 (Decimal('116.78'), 'Nepal'),
 (Decimal('523.79'), 'Saudi Arabia'),
 (Decimal('73.76'), 'Lithuania'),
 (Decimal('402.05'), 'Bangladesh'),
 (Decimal('259.43'), 'Yugoslavia'),
 (Decimal('586.66'), 'Netherlands'),
 (Decimal('129.70'), 'Gambia'),
 (Decimal('1511.48'), 'Nigeria'),
 (Decimal('318.23'), 'Dominican Republic'),
 (Decimal('694.39'), 'Egypt'),
 (Decimal('114.72'), 'Faroe Islands'),
 (Decimal('134.67'), 'Zambia'),
 (Decimal('4110.32'), 'United States'),
 (Decimal('1434.48'), 'Argentina'),
 (Decimal('89.74'), 'Slovakia'),
 (Decimal('215.48'), 'Angola'),
 (Decimal('187.50'), 'Oman'),
 (Decimal('3045.87'), 'Russian Federation'),
 (Decimal('136.73'), 'Turkmenistan'),
 (Decimal('67.82'), 'Afghanistan'),
 (Decimal('133.71'), 'Czech Republic'),
 (Decimal('114.72'), 'Liechtenstein'),
 (Decimal('91.77'), 'Ethiopia'),
 (Decimal('245.43'), 'Azerbaijan'),
 (Decimal('3200.52'), 'Brazil'),
 (Decimal('422.01'), 'Israel'),
 (Decimal('100.75'), 'Senegal'),
 (Decimal('328.29'), 'Chile'),
 (Decimal('92.76'), 'New Zealand'),
 (Decimal('111.71'), 'Hungary'),
 (Decimal('344.20'), 'Mozambique'),
 (Decimal('3307.04'), 'Mexico'),
 (Decimal('101.74'), 'Finland'),
 (Decimal('1210.94'), 'Taiwan'),
 (Decimal('116.70'), 'Sri Lanka'),
 (Decimal('419.04'), 'Thailand'),
 (Decimal('115.70'), 'Estonia'),
 (Decimal('262.40'), 'Latvia'),
 (Decimal('468.88'), 'Peru'),
 (Decimal('683.30'), 'Venezuela'),
 (Decimal('216.54'), 'Runion'),
 (Decimal('152.66'), 'Holy See (Vatican City State)'),
 (Decimal('106.65'), 'Anguilla'),
 (Decimal('112.75'), 'Bahrain'),
 (Decimal('383.10'), 'Algeria'),
 (Decimal('369.15'), 'Malaysia'),
 (Decimal('950.75'), 'Iran'),
 (Decimal('606.58'), 'Spain'),
 (Decimal('730.42'), 'Ukraine'),
 (Decimal('574.65'), 'South Korea'),
 (Decimal('277.34'), 'Belarus'),
 (Decimal('73.82'), 'Tonga'),
 (Decimal('103.78'), 'French Guiana'),
 (Decimal('142.70'), 'Hong Kong'),
 (Decimal('123.72'), 'Malawi'),
 (Decimal('148.69'), 'Nauru'),
 (Decimal('235.38'), 'Romania'),
 (Decimal('113.69'), 'North Korea'),
 (Decimal('254.39'), 'Puerto Rico'),
 (Decimal('593.63'), 'Canada'),
 (Decimal('5802.73'), 'China'),
 (Decimal('307.22'), 'Austria'),
 (Decimal('127.66'), 'Moldova'),
 (Decimal('1204.15'), 'South Africa'),
 (Decimal('253.46'), 'Kenya'),
 (Decimal('877.97'), 'Poland'),
 (Decimal('198.48'), 'Kazakstan'),
 (Decimal('924.80'), 'United Kingdom'),
 (Decimal('113.65'), 'Brunei'),
 (Decimal('78.77'), 'Tunisia'),
 (Decimal('831.11'), 'Italy'),
 (Decimal('525.72'), 'Pakistan'),
 (Decimal('144.66'), 'Sweden'),
 (Decimal('118.75'), 'Armenia'),
 (Decimal('232.46'), 'Greece'),
 (Decimal('6630.27'), 'India'),
 (Decimal('510.83'), 'Yemen'),
 (Decimal('275.38'), 'Paraguay'),
 (Decimal('194.48'), 'Myanmar'),
 (Decimal('2381.32'), 'Philippines'),
 (Decimal('307.29'), 'Morocco'),
 (Decimal('137.66'), 'Greenland'),
 (Decimal('252.39'), 'Switzerland'),
 (Decimal('1510.33'), 'Indonesia'),
 (Decimal('120.74'), 'Tuvalu'),
 (Decimal('183.53'), 'Bolivia'),
 (Decimal('96.75'), 'Saint Vincent and the Grenadines'),
 (Decimal('212.50'), 'Congo, The Democratic Republic of the'),
 (Decimal('227.46'), 'Sudan'),
 (Decimal('333.16'), 'United Arab Emirates'),
 (Decimal('341.17'), 'Tanzania'),
 (Decimal('204.50'), 'Bulgaria'),
 (Decimal('235.46'), 'French Polynesia')]

## Revenue by Month
TODO: Write a query that calculates total revenue (sales_amount) by month

In [7]:
%%sql
SELECT sum(sales_amount) as revenue, month
FROM factsales f join dimdate d on f.date_key=d.date_key group by month;

 * postgresql://student:***@127.0.0.1:5432/pagila_star
5 rows affected.


[(Decimal('4824.43'), 1),
 (Decimal('23886.56'), 3),
 (Decimal('28559.46'), 4),
 (Decimal('9631.88'), 2),
 (Decimal('514.18'), 5)]

## Revenue by Month & Country
TODO: Write a query that calculates total revenue (sales_amount) by month and country. Sort the data by month, country, and revenue in descending order. The first few rows of your output should match the table below.

In [17]:
%%sql
SELECT month,country,sum(sales_amount) as revenue
FROM factsales f 
join dimdate d on d.date_key=f.date_key
join dimstore c on c.store_key=f.store_key group by month, country order by month,country,revenue desc

 * postgresql://student:***@127.0.0.1:5432/pagila_star
10 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
</tbody></table></div>

## Revenue Total, by Month, by Country, by Month & Country All in one shot

TODO: Write a query that calculates total revenue at the various grouping levels done above (total, by month, by country, by month & country) all at once using the grouping sets function. Your output should match the table below.

In [19]:
%%sql
SELECT month,country,sum(sales_amount) as revenue
FROM factsales f 
join dimdate d on d.date_key=f.date_key
join dimstore c on c.store_key=f.store_key
group by grouping sets ((),month,country,(month,country))

 * postgresql://student:***@127.0.0.1:5432/pagila_star
18 rows affected.


[(1, 'Australia', Decimal('2364.19')),
 (1, 'Canada', Decimal('2460.24')),
 (1, None, Decimal('4824.43')),
 (2, 'Australia', Decimal('4895.10')),
 (2, 'Canada', Decimal('4736.78')),
 (2, None, Decimal('9631.88')),
 (3, 'Australia', Decimal('12060.33')),
 (3, 'Canada', Decimal('11826.23')),
 (3, None, Decimal('23886.56')),
 (4, 'Australia', Decimal('14136.07')),
 (4, 'Canada', Decimal('14423.39')),
 (4, None, Decimal('28559.46')),
 (5, 'Australia', Decimal('271.08')),
 (5, 'Canada', Decimal('243.10')),
 (5, None, Decimal('514.18')),
 (None, None, Decimal('67416.51')),
 (None, 'Australia', Decimal('33726.77')),
 (None, 'Canada', Decimal('33689.74'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>month</th>
        <th>country</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>1</td>
        <td>Australia</td>
        <td>2364.19</td>
    </tr>
    <tr>
        <td>1</td>
        <td>Canada</td>
        <td>2460.24</td>
    </tr>
    <tr>
        <td>1</td>
        <td>None</td>
        <td>4824.43</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Australia</td>
        <td>4895.10</td>
    </tr>
    <tr>
        <td>2</td>
        <td>Canada</td>
        <td>4736.78</td>
    </tr>
    <tr>
        <td>2</td>
        <td>None</td>
        <td>9631.88</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Australia</td>
        <td>12060.33</td>
    </tr>
    <tr>
        <td>3</td>
        <td>Canada</td>
        <td>11826.23</td>
    </tr>
    <tr>
        <td>3</td>
        <td>None</td>
        <td>23886.56</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Australia</td>
        <td>14136.07</td>
    </tr>
    <tr>
        <td>4</td>
        <td>Canada</td>
        <td>14423.39</td>
    </tr>
    <tr>
        <td>4</td>
        <td>None</td>
        <td>28559.46</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Australia</td>
        <td>271.08</td>
    </tr>
    <tr>
        <td>5</td>
        <td>Canada</td>
        <td>243.10</td>
    </tr>
    <tr>
        <td>5</td>
        <td>None</td>
        <td>514.18</td>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>67416.51</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Australia</td>
        <td>33726.77</td>
    </tr>
    <tr>
        <td>None</td>
        <td>Canada</td>
        <td>33689.74</td>
    </tr>
</tbody></table></div>